File: pyround1/530.二叉搜索树的最小绝对差.py
#
# @lc app=leetcode.cn id=530 lang=python
#
# [530] 二叉搜索树的最小绝对差
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def getMinimumDifference(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        st = []
        p = root
        pre = -float('inf')
        min_val = float('inf')
        while p is not None or st:
            while p is not None:
                st.append(p)
                p = p.left
            p = st.pop()
            cur = p.val
            if cur - pre < min_val:
                min_val = cur - pre
            pre = cur
            p = p.right
        return min_val
# @lc code=end

File: 530.二叉搜索树的最小绝对差.cpp
/*
 * @lc app=leetcode.cn id=530 lang=cpp
 *
 * [530] 二叉搜索树的最小绝对差
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> res;
    int getMinimumDifference(TreeNode* root) {
        helper(root);
        int min_val = INT_MAX;
        for(int i = 1; i < res.size(); i++){
            min_val = min(min_val, res[i] - res[i-1]);
        }
        return min_val;
    }
    void helper(TreeNode* root ){
        if (!root)
            return ;
        if(root->left)
        {
            helper(root->left);
        }
        res.push_back(root->val);
        if(root->right)
        {
            helper(root->right);
        }
    }
};
// @lc code=end

